### ISSUE TITLE
Delays getting client accepted [troubleshooting advice needed]

### ISSUE THREAD

**Issue by localstorm** [netty/netty#9248]
Hi,

My system is built using two Netty-based applications where one is serving inbound requests coming at high rate and occasionally speaking to the second one for extra data. What we have seen although rather rarely (which makes it difficult to repro) is the following:

First application is getting steady high request rate and everything is working fine. Second application is getting requests from the First one at significantly lower rate (let's say 100x less requests). However, once in a while we see that First application times out after configured 2 seconds without getting anything from Second one and the Second application shows this client accepted (Channel pipeline created) already much after (5-7 seconds later) First app timed out. 

My immediate theory was that the handler thread in the Second application was blocked on something and didn't get to accepting new clients, but after adding logs we have a prof that the worker thread which ended up getting connection in Second app was idle since before First app decided to connect. To be more precise the proof is that it dind't do anything for the app, so it was not in the handler's code.

So basically from the point of view of the Second app client got connected and even sent a request, but dropped connection before Second app was able to respond, but from the p.o.v. of the First app it connected, sent a request, but timed out without getting any response a long time before Second one get the connection.

The average load on the server is rather low. 

What would you suggest to try here to understand the problem better? My thinking is maybe split threads belonging to Second app into parent pool and child pool and have fewer threads in parent, but running at higher priority? Not sure otherwise what it might be if not an OS thread scheduling issue.

Any advice is highly appreciated.

**Comment by localstorm** [comment]
What is also interesting here is that when it happens other threads of the Second application are also likely to show the same issue. So we hit this in batches of 5-10 connections at approximately the same time. They see client coming and going away after receiving a few request bytes a long after actual client gave up on them

**Comment by DATuan91** [comment]
Can you show the config of server and client bootstrap @localstorm.?
Anyway, this issue is rather vague, but you could try do the following: 

- Add an interval ping/ polling mechanic to keep connection alive
- Try tweaking ChannelOption.SO_KEEPALIVE, ChannelOption.ALLOW_HALF_CLOSURE, ChannelOption.SO_LINGER or ChannelOption.TCP_NODELAY. Maybe other config from io.netty.channel.ChannelOption might help with your handler pipeline, too
- You can change the log level to TRACE and add io.netty.handler.logging.LoggingHandler to handler pipeline

**Comment by localstorm** [comment]
Ops, missed your response. We don't have a connection to maintain with keepalive or pings. Connection is closed every time server responds. But sometimes server doesn't get ChannelPipeline created before client times out, which is something we still don't understand. 

Unfortunately, I don't think TRACE is an option here, we'll fill our disks very quickly.

**Do you think we might want to try playing with parent workers being a separate pool of threads than child workers?**

Here is the client Bootstrap:

bootstrap.group(loop)
                    .channel(EnvironmentUtils.chooseSocketClass())
                    .resolver(NoDnsResolverGroup.INSTANCE)         // We don't event want a possibility of a DNS call
                    .option(ChannelOption.ALLOCATOR,               PooledByteBufAllocator.DEFAULT)
                    .option(ChannelOption.ALLOW_HALF_CLOSURE,      false)
                    .option(ChannelOption.SO_KEEPALIVE,            true)
                    .option(ChannelOption.CONNECT_TIMEOUT_MILLIS,  5000)
                    .option(ChannelOption.WRITE_BUFFER_WATER_MARK, WRITE_BUFFER_WATER_MARK)
                    .option(ChannelOption.SO_RCVBUF,  256 * 1024)
                    .option(ChannelOption.RCVBUF_ALLOCATOR,        BuffersConstants.COMM_BUFFER_ALLOCATOR);

This is the server:

serverBootstrap.group(workers)
                .channel(EnvironmentUtils.chooseServerSocketClass())
                .childHandler(dataStreamChannelInitializer)
                .option(ChannelOption.ALLOCATOR,                PooledByteBufAllocator.DEFAULT)
                .option(ChannelOption.SO_REUSEADDR,             true)
                .option(ChannelOption.SO_BACKLOG,               128)
                .option(ChannelOption.SO_RCVBUF, 64*1024)
                .childOption(ChannelOption.AUTO_READ,           true)
                .childOption(ChannelOption.ALLOW_HALF_CLOSURE,  false)
                .childOption(ChannelOption.SO_KEEPALIVE,        true)
.childOption(ChannelOption.RCVBUF_ALLOCATOR, BuffersConstants.COMM_BUFFER_ALLOCATOR);

**Comment by localstorm** [comment]
Actually, maybe TRACE is possible, I'll come back with what I found.

**Comment by normanmaurer** [comment]
Also you may want to attach profiler to see if you block the EventLoop

**Comment by localstorm** [comment]
We have nailed it.

So, the long story short, it doesn't have anything to do with Netty per se, but it's interesting enough, so I wanted to share it here.

So we traced these huge pauses back to JVM parameter -Xlog:safepoint. What happens there is that this parameter was making JVM print to STDOUT all GC-related info. On a surface of it, it doesn't cause any issue, but in our prod stacks we have STDOUT redirected to a file, which is expected to stay empty or have no traffic there. We do that to capture something if it bombs out in unexpected place like unhandled exceptions or OOM error.

Now, what happens with GC output printed to file is that occasionally, especially in a cloud environment riddled with network-attached storage, these writes suspend the whole JVM and apparently, when this happens, GC cannot start while IO call is stuck. More about this here: https://engineering.linkedin.com/blog/2016/02/eliminating-large-jvm-gc-pauses-caused-by-background-io-traffic

With that flag removed, we saw a complete elimination of the problem.

Cheers,
Alexey
